### ISSUE TITLE
Callback protocol with TypeVarTuple and keyword-only argument doesn't type check

### ISSUE THREAD

**Issue by chris-chambers** [microsoft/pyright#2948]
**Describe the bug**
Pyright is unable to type check a callback protocol that accepts a variable number of positional arguments (given by an unpacked TypeVarTuple) followed by keyword-only arguments.

**To Reproduce**
Define such a callback protocol and then try to assign a function that conforms to the protocol to an explicitly-hinted variable (see attached code).

**Expected behavior**
The assignment should type check without error.  According to my understanding of [PEP-646](https://www.python.org/dev/peps/pep-0646/), it should be possible to define this kind of callback protocol.  A [similar example](https://www.python.org/dev/peps/pep-0646/#args-as-a-type-variable-tuple) is given in PEP-646 when discussing prefixes and suffixes to unpacked tuple types:

> As usual, we can unpack any tuple types. For example, by using a type variable tuple inside a tuple of other types, we can refer to prefixes or suffixes of the variadic argument list. For example:
>
> ```python
> # os.execle takes arguments 'path, arg0, arg1, ..., env'
> def execle(path: str, *args: *Tuple[*Ts, Env]) -> None: ...
> ```
>
> Note that this is different to
>
> ```python
> def execle(path: str, *args: *Ts, env: Env) -> None: ...
> ```
> as **this would make env a keyword-only argument.** *[emphasis added]*

**Screenshots or Code**
```python
# pyright: strict

from typing import Protocol, TypeVar
from typing_extensions import TypeVarTuple, Unpack

T = TypeVar("T")
Ts = TypeVarTuple("Ts")


class CallbackA(Protocol[Unpack[Ts], T]):
    # The presence of *args: *Ts should begin the keyword-only section of the
    # parameter list, but `keyed` is not treated as keyword-only.
    def __call__(self, *args: *Ts, keyed: T) -> tuple[Unpack[Ts], T]:
        ...


class CallbackB(Protocol[Unpack[Ts], T]):
    # The extra * marker is a syntax error, since *args: *Ts is already present,
    # but it does get pyright to treat `keyed` as keyword-only.
    def __call__(self, *args: *Ts, *, keyed: T) -> tuple[Unpack[Ts], T]:
        ...


def example(a: int, b: str, *, keyed: bool) -> tuple[int, str, bool]:
    return (a, b, keyed)


a: CallbackA[int, str, bool] = example
b: CallbackB[int, str, bool] = example

reveal_type(a)
reveal_type(b)
```

Here is the output of pyright:
```
No configuration file found.
No pyproject.toml file found.
stubPath /Users/chris/projects/pyright-bugs/1/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Auto-excluding /Users/chris/projects/pyright-bugs/1/venv
Found 1 source file
/Users/chris/projects/pyright-bugs/1/example.py
  /Users/chris/projects/pyright-bugs/1/example.py:20:36 - error: Keyword-only argument separator not allowed after "*" parameter
  /Users/chris/projects/pyright-bugs/1/example.py:10:7 - error: Type variable "T" used in generic protocol "CallbackA" should be covariant (reportInvalidTypeVarUse)
  /Users/chris/projects/pyright-bugs/1/example.py:17:7 - error: Type variable "T" used in generic protocol "CallbackB" should be covariant (reportInvalidTypeVarUse)
  /Users/chris/projects/pyright-bugs/1/example.py:28:32 - error: Expression of type "(a: int, b: str, *, keyed: bool) -> tuple[int, str, bool]" cannot be assigned to declared type "CallbackA[int, str, bool]"
    Type "(a: int, b: str, *, keyed: bool) -> tuple[int, str, bool]" cannot be assigned to type "(int, str, keyed: bool) -> tuple[int, str, bool]"
      Function accepts too many positional parameters; expected 2 but received 3
        Keyword parameter "keyed" is missing in destination (reportGeneralTypeIssues)
  /Users/chris/projects/pyright-bugs/1/example.py:31:13 - information: Type of "a" is "CallbackA[int, str, bool]"
  /Users/chris/projects/pyright-bugs/1/example.py:32:13 - information: Type of "b" is "(a: int, b: str, *, keyed: bool) -> tuple[int, str, bool]"
4 errors, 0 warnings, 2 informations 
Completed in 0.527sec
```

The errors about variance are also incorrect (T is used an both an input and output type, so it should be invariant (or really, should be two TypeVars with a bound relationship that can't currently be expressed in Python)).  But, I think there is a fair chance the variance error is a false positive being caused by whatever causes the function signature problem.

**VS Code extension or command-line**
Present in both
- Pyright 1.1.215
- Pylance 2022.1.5

**Comment by erictraut** [comment]
Thanks for the great bug report. This code sample actually uncovered three separate bugs. All of them will be fixed in the next release. I've added a unit test based on your example.

* Fixed bug in callable type compatibility logic. It was not properly handling some edge cases where a keyword parameter in the source and destination had an incompatible type if one or both of the types were specialized.
* Fixed a bug that resulted in false positives when specializing a callback protocol with a TypeVarTuple when the callback protocol also contained one or more keyword arguments that were generic. The logic needed to insert a keyword-only separator into the signature after the unpacked TypeVarTuple parameters.
* Fixed a bug in the protocol invariance checking logic. It wasn't properly handling protocols that used a TypeVarTuple as a type parameter that wasn't last (right-most) in the type parameter list.

I'm glad to see that someone is trying out the PEP 646 support. This PEP added a significant amount of complexity to the type checking logic, so there are bound to be some edge cases that I missed when implementing it. Thanks for your patience in helping me work through any issues.

**Comment by erictraut** [comment]
This is included in pyright 1.1.216, which I just published. It will also be included in the next release of pylance.

**Comment by chris-chambers** [comment]
Thank you for the impossibly fast fix!
